keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,354 @@
+    * [Inline CSI volume operation stages](#inline-csi-volume-operation-stages)
+    * [Security considerations](#security-considerations)
+
+## Summary
+Currently, volumes that are backed by CSI drivers can only be used with the `PersistentVolume` and `PersistentVolumeClaim` objects. This proposal is to implement support for the ability to nest CSI volume declarations within pod specs, as is supported already by existing in-tree storage providers.
+
+This KEP started life as [feature #2273](https://github.com/kubernetes/community/pull/2273).  Please follow that link for historical context.
+
+
+## Motivation
+Implementing support for embedding volumes directly in pod specs would allow external CSI drivers to have feature parity with their in-tree counter parts.  There are several reasons why this is desirable:
+* It provides the community a path to move away from in-tree volume plugins to CSI, as designed in a separate proposal https://github.com/kubernetes/community/pull/2199/. 
+* This feature would make it possible to create new types of CSI drivers such as ephemeral volume drivers.  They can be used to inject arbitrary states such as configuration, secrets, identity, variables or similar information inside pods. 
+* This can also provide a migration path for older Flex-style drivers (allowing the deprecation of Flex.)
+
+### Goals 
+* Provide a high level design for inline CSI volumes support
+* Define API changes needed to support this feature
+* Inline CSI volumes should work with existing CSI drivers
+* Design for ephemeral inline CSI volumes support
+* Ensure that inline CSI volumes usage is secure
+
+### Non-goals
+The followings will not be addressed by this KEP:
+* Introduce new CSI spec changes to support this feature
+* Introduce required changes to existing CSI drivers for this feature
+* Implementation of ephemeral inline support
+* Support for topology or pod placement scheme for ephemeral inline volumes
+* Support for PV/PVC originated volumes using inline
+* Support for PV/PVC related features such as topology, raw block, mount options, and resizing
+
+## User stories
+* As a storage provider, I want to be able to create CSI drivers that support persistent volumes that can be nested within pod specs.  These volumes would work similarly to how my current in-tree drivers work (with minor limitations addressed later).
+* As a storage provider, I want to use the CSI API to develop arbitrary drivers that can mount ephemeral volumes which follow the lifecycles of pods where they are embedded.   These ephemeral volumes do not follow the full CSI storage workflow (of provision, stage, attach, mount, unmount, detach, unstage, delete)
+* As a user, I want to be able to deploy pods, with persistent CSI volumes embedded in their specs, without the use of PV/PVC.  
+* As a user I want to be able to define pod specs with embedded ephemeral CSI volumes that are created/mounted when the pod is deployed and are deleted when the pod goes away.
```

Comment:
It would help the reader to know what the use cases are for out of tree ephemeral CSI volumes.  From a user story perspective, I'm missing the "why", or a reference/link to the why.

Reply:
Adding to the storage provider user story, why ephemeral CSI drivers would be useful.  Thanks.